### ISSUE TITLE
Unknown shaped and negative loss

### ISSUE THREAD

**Issue by tvatsal1996** [HawkAaron/warp-transducer#65]
Hi :
I am using the GPU version of tensorflow binding warprnnt_tensorflow.rnnt_loss(), but the loss has \<unknown\> shape and negative loss is also encountered.
Shapes of target and logits is as follows:
targets: [batch_size, U] -> (?, ?)
logits: [batch_size, T, U+1, hidden] -> (?, ?, ?, hidden)
input_len: [batch_size,] -> (?,)
target_len: [batch_size,] -> (?,)
warprnnt_tensorflow.rnnt_loss(logits, targets, input_len, target_len): \<unknown\>
Why loss shape is not [batch_size,] -> (?,)
Sometimes this loss comes out as negative also during training.

**Comment by BuaaAlban** [comment]
> Hi :
> I am using the GPU version of tensorflow binding warprnnt_tensorflow.rnnt_loss(), but the loss has <unknown> shape and negative loss is also encountered.
> Shapes of target and logits is as follows:
> targets: [batch_size, U] -> (?, ?)
> logits: [batch_size, T, U+1, hidden] -> (?, ?, ?, hidden)
> input_len: [batch_size,] -> (?,)
> target_len: [batch_size,] -> (?,)
> warprnnt_tensorflow.rnnt_loss(logits, targets, input_len, target_len): <unknown>
> Why loss shape is not [batch_size,] -> (?,)
> Sometimes this loss comes out as negative also during training.

I also have the question that is it reasonable to have negative loss, maybe it is because the probability i s the sum of all alignments, which means the probability  can be larger than 1 and the loss can be negative?

**Comment by tvatsal1996** [comment]
> > Hi :
> > I am using the GPU version of tensorflow binding warprnnt_tensorflow.rnnt_loss(), but the loss has  shape and negative loss is also encountered.
> > Shapes of target and logits is as follows:
> > targets: [batch_size, U] -> (?, ?)
> > logits: [batch_size, T, U+1, hidden] -> (?, ?, ?, hidden)
> > input_len: [batch_size,] -> (?,)
> > target_len: [batch_size,] -> (?,)
> > warprnnt_tensorflow.rnnt_loss(logits, targets, input_len, target_len): 
> > Why loss shape is not [batch_size,] -> (?,)
> > Sometimes this loss comes out as negative also during training.
> 
> I also have the question that is it reasonable to have negative loss, maybe it is because the probability i s the sum of all alignments, which means the probability can be larger than 1 and the loss can be negative?

Negative loss should not be anything meaningful. It might be happening due to wrong kind of input as logits. In the meanwhile, I have another question, why we are doing log-softmax and not simply softmax to the logits? Log-softmax will bring the logits in range (-inf, 0). Doing softmax will give us the probability distribution.

**Comment by HawkAaron** [comment]
The negative loss bug may have been fixed in the most recent commit.

@tvatsal1996  As https://github.com/HawkAaron/warp-transducer/blob/master/README.md#interface shown,  for gpu version, activations before softmax should be the input.

**Comment by tvatsal1996** [comment]
> The negative loss bug may have been fixed in the most recent commit.
> 
> @tvatsal1996 As https://github.com/HawkAaron/warp-transducer/blob/master/README.md#interface shown, for gpu version, activations before softmax should be the input.

Hi, I found negative loss was due to the issue that I did not append 0 to the input of prediction encoder, so my dimension of the final 4d acts were (B,T,U,V) instead of (B,T,U+1,V). After fixing this, I stopped getting negative values. Yes, I was following your instructions regarding CPU and GPU implementation, but the error was from my end. Thanks for your response :)

**Comment by BuaaAlban** [comment]
> The negative loss bug may have been fixed in the most recent commit.
> 
> @tvatsal1996 As https://github.com/HawkAaron/warp-transducer/blob/master/README.md#interface shown, for gpu version, activations before softmax should be the input.

I also find my mapping for timit 60-48-39 was wrong, the data label is in 48, and my mapping for encoding is 60-39 instead of 48-39，I think this should be the reason of negative loss
